Failed to convert between Type and SumType

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.4 LTS (WSL 2)
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
CC version: cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

getwd: /mnt/d/Programming/V/playground
vmodules: /home/caleb/.vmodules
vroot: /mnt/d/Programming/V/src
vexe: /mnt/d/Programming/V/src/v
vexe mtime: 2022-03-01 14:11:34
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 3b6e122.996bd41

Git version: git version 2.25.1
Git vroot status: weekly.2022.09-3-g996bd41c
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg playground.v`
```v
struct Point {
	x int
	y int
}

struct Line {
	p1 Point
	p2 Point
}

// Sum type
type ObjSumType = Line | Point

fn main() {
	// Type checking and casts
	mut point := ObjSumType(Point{2, 5})

	if point is Point {
		println('Point')
	}

	if point !is Point {
		println('Not Point')
	}

	if mut point is Point {
		println(point)
	}
}
```

**What did you expect to see?**

A println based on the type used


**What did you see instead?**
```
==================
/tmp/v_1000/playground.13397920522120289165.tmp.c:11383: error: cannot convert 'struct main__Point' to 'struct main__ObjSumType'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -910,12 +910,13 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 					g.gen_expr_to_string(node.left, rec_type)
 					return
 				} else if node.left.obj.smartcasts.len > 0 {
-					cast_sym := g.table.sym(node.left.obj.smartcasts.last())
+					rec_type = node.left.obj.smartcasts.last()
+					cast_sym := g.table.sym(rec_type)
 					if cast_sym.info is ast.Aggregate {
 						rec_type = cast_sym.info.types[g.aggregate_type_idx]
-						g.gen_expr_to_string(node.left, rec_type)
-						return
 					}
+					g.gen_expr_to_string(node.left, rec_type)
+					return
 				}
 			}
 		}
@@ -1297,7 +1298,8 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 					if expr.obj is ast.Var {
 						typ = expr.obj.typ
 						if expr.obj.smartcasts.len > 0 {
-							cast_sym := g.table.sym(expr.obj.smartcasts.last())
+							typ = expr.obj.smartcasts.last()
+							cast_sym := g.table.sym(typ)
 							if cast_sym.info is ast.Aggregate {
 								typ = cast_sym.info.types[g.aggregate_type_idx]
 							}

--- a/vlib/v/gen/c/str_intp.v
+++ b/vlib/v/gen/c/str_intp.v
@@ -156,7 +156,8 @@ fn (mut g Gen) str_val(node ast.StringInterLiteral, i int) {
 				if g.comptime_var_type_map.len > 0 || g.comptime_for_method.len > 0 {
 					exp_typ = expr.obj.typ
 				} else if expr.obj.smartcasts.len > 0 {
-					cast_sym := g.table.sym(expr.obj.smartcasts.last())
+					exp_typ = expr.obj.smartcasts.last()
+					cast_sym := g.table.sym(exp_typ)
 					if cast_sym.info is ast.Aggregate {
 						exp_typ = cast_sym.info.types[g.aggregate_type_idx]
 					}

--- a/vlib/v/tests/print_smartcast_variable_test.v
+++ b/vlib/v/tests/print_smartcast_variable_test.v
@@ -1,0 +1,33 @@
+struct Point {
+	x int
+	y int
+}
+
+struct Line {
+	p1 Point
+	p2 Point
+}
+
+// Sum type
+type ObjSumType = Line | Point
+
+fn test_print_smartcast_variable() {
+	// Type checking and casts
+	mut point := ObjSumType(Point{2, 5})
+
+	if point is Point {
+		println('Point')
+	}
+
+	if point !is Point {
+		println('Not Point')
+	}
+
+	if mut point is Point {
+		println(point)
+		assert point.str().contains('x: 2')
+		assert point.str().contains('y: 5')
+		assert '$point'.contains('x: 2')
+		assert '$point'.contains('y: 5')
+	}
+}

